Convert the entire codebase to TypeScript

mstange opened the pull request:
FIxes #2931.

I squashed and re-split the changes from #5531.

`yarn test-all` passes after every commit.

[702 files changed, 13512 insertions(+), 64132 deletions(-)](https://gist.github.com/mstange/ed98002f852ba6f3e5979bae9cc6571c)

canova commented on package.json:
```diff
@@ -29,11 +29,7 @@
-    "flow-stop": "flow stop",
-    "flow-coverage": "npx flow-coverage-report -i 'src/**/*.js' -t html -t text",
-    "flow-generate-libdefs": "npx flow-typed install --libdefDir src/types/libdef",
+    "typecheck": "tsc --noEmit",
```
nit: can we maybe name this `ts`? `yarn ts` is shorter and easier to type for me. And we already have the same command in [profiler-server](https://github.com/firefox-devtools/profiler-server/blob/2de28e818a1214d1af9fc3e9ea2e7a76f4ba1697/package.json#L46).

mstange replied:
Absolutely!

canova reviewed (commented):
Oh wow, this PR is huge, amazing work! I couldn't finish the review yet, I'm at ecff67d11b2cf2e94b3e0a36db0c07e67e47d2ca (Convert window-console.js and index.js.) at the moment. At least all the shipping code is done! Tomorrow I will continue with the rest. I'm submitting the comments that I wrote so far, so you can look at my them sooner than later. I expect to finish the review tomorrow.

I see a few of places where I see flow being mentioned. But I think they can be removed as a follow-ups.

For example:
- We still have `utils/flow.ts`. It's funny that it's flow.ts now, maybe we can change it to utils/typescript.ts or utils/type.ts
- I see some flow mentions inside src/reducers/README.md
- Some more flow mentions in the comments.

But they all could be done as follow-ups.

Some files for some reason don't appear they are `git mv`ed even though they are in the same commit, but that's fine I think.

Also as a follow-up, it might be good to add the commit hashes of these inside `.git-blame-ignore-revs` so they would be ignored from the blame? (don't know if adding the merge commit hash is enough or we need to add every one of them individually)

Ah lastly, can you run `npx yarn-deduplicate` to deduplicate some of the packages that we added?

canova reviewed (approved):
I'm done going through the commits! It was easier to look at the second half.

Wow, thanks a lot for the migration! Looks great. Let's land this and we can follow-up with the improvements.

FWIW, I was using this branch locally to test stuff too and I can really feel the difference between flow and this! It was taking a long time to get LSP errors on my editor with flow, now I'm getting near instant feedback on errors! The experience difference is night and day. And after this, I can finally remove every config/plugin related to Flow from my editor.

mstange commented on package.json:
```diff
@@ -29,12 +29,8 @@
-    "flow-coverage": "npx flow-coverage-report -i 'src/**/*.js' -t html -t text",
-    "flow-generate-libdefs": "npx flow-typed install --libdefDir src/types/libdef",
-    "protoc": "npx -p protobufjs-cli pbjs -t static-module -w commonjs -o ./src/profile-logic/import/proto/simpleperf_report.js ./src/profile-logic/import/proto/simpleperf_report.proto",
+    "ts": "tsc --noEmit",
```
Ah, yes.

mstange commented on tsconfig.json:
```diff
@@ -0,0 +1,47 @@
+
+    // Module Resolution
+    "resolveJsonModule": true,
+    "baseUrl": ".",
```
seems not, I'll remove it.

mstange commented on README.md:
````diff
@@ -41,7 +41,7 @@ yarn install
 yarn start
 ```
 
-[Flow](https://flow.org/) is used for type checking. VSCode users can install the ["Flow Language Support" extension](https://marketplace.visualstudio.com/items?itemName=flowtype.flow-for-vscode), and disable VSCode's built-in TypeScript extension in the workspace via the [setup instructions here](https://github.com/flow/flow-for-vscode#setup).
````
fixed

mstange commented on src/types/indexeddb.ts:
oops, yes removed. We may want to revive it at some point, I just didn't know how to not make it conflict with the built-in IndexedDB types. Julien wrote it because it caught more mistakes than the built-in types, because it was checking the index and value types too.

mstange commented on src/actions/app.ts:
```diff
@@ -63,7 +62,7 @@ export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
       });
       dispatch({
-        type: 'CHANGE_SELECTED_TAB',
+        type: 'CHANGE_SELECTED_TAB' as const,
```
I've removed the `as const`s.

mstange commented:
Thanks for the thorough review of this huge PR!


> I see a few of places where I see flow being mentioned. But I think they can be removed as a follow-ups.

I ended up fixing them all.

> For example:
> 
>     * We still have `utils/flow.ts`. It's funny that it's flow.ts now, maybe we can change it to utils/typescript.ts or utils/type.ts

Renamed to `utils/types.ts`. I don't like the file very much - I think files under `utils` should be the "leaves" of the module dependency tree, and not depend on type definitions for our actions etc. I think functions like `toValidTabSlug` should be close to where we define the `TabSlug` type, etc.

>     * I see some flow mentions inside src/reducers/README.md

Fixed, and I went through the other .md files as well.

>     * Some more flow mentions in the comments.

Oh actually this one I haven't fixed.

> Some files for some reason don't appear they are `git mv`ed even though they are in the same commit, but that's fine I think.

Yeah I'm not sure why. Maybe because those didn't hit the percentage threshold of preserved lines.

> Also as a follow-up, it might be good to add the commit hashes of these inside `.git-blame-ignore-revs` so they would be ignored from the blame? (don't know if adding the merge commit hash is enough or we need to add every one of them individually)

Ah yes, good idea.

> Ah lastly, can you run `npx yarn-deduplicate` to deduplicate some of the packages that we added?

Done.

> FWIW, I was using this branch locally to test stuff too and I can really feel the difference between flow and this! It was taking a long time to get LSP errors on my editor with flow, now I'm getting near instant feedback on errors! The experience difference is night and day. And after this, I can finally remove every config/plugin related to Flow from my editor.

🎉
